rbd: add validation to ToCSI() for rbdVolume and rbdSnapshot

nixpanic opened the pull request:
After an unfortnate timed restart of the provisioner, a volume that got
cloned did not get a `rbdVolume.VolID` set. The `.VolID` is used as the
CSI Volume Handle, and is a required attribute.

The `rbdVolume` and `rbdSnapshot` structs have a `.ToCSI()` function
that can do the validation of required attributes. This is now added,
including unit-tests.

nixpanic committed:
rbd: add validation to ToCSI() for rbdVolume and rbdSnapshot

After an unfortnate timed restart of the provisioner, a volume that got
cloned did not get a `rbdVolume.VolID` set. The `.VolID` is used as the
CSI Volume Handle, and is a required attribute.

The `rbdVolume` and `rbdSnapshot` structs have a `.ToCSI()` function
that can do the validation of required attributes. This is now added,
including unit-tests.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -244,6 +244,17 @@ func (cs *ControllerServer) parseVolCreateRequest(
 func (rbdVol *rbdVolume) ToCSI(ctx context.Context) (*csi.Volume, error) {
+	switch {
+	case rbdVol.VolID == "":
+		return nil, fmt.Errorf("%q may not have an empty volume-id", rbdVol)
```
need to replace the error statement? because it says `may not`  it should be `may have`

nixpanic replied:
Well, the error is that the rbdVolume has an empty volume-id, and that is not allowed. Do you want it like this?
```go
fmt.Errorf("%q has an empty volume-id", rbdVol)
```
To me the meaning is the same, but if the 2nd is clearer, I'm happy to change it.

Madhu-1 replied:
make sense, but `may not have an empty volume-id` should be `may have an empty volume-id` we should remove `not` from it , isnt it?

nixpanic replied:
The error would read like `replicated/image may not have an empty volume-id` when the volume-id empty.

I think it is strange to report an error like `replicated/image may have an empty volume-id` when the volume-id is, in fact, empty. So, it actually _may not have_ an empty volume-id (why return an error?).

Maybe this is better?
```go
fmt.Errorf("%q does not have a volume-id set", rbdVol)
```

Madhu-1 replied:
sounds good 👍🏻

nixpanic replied:
done, please check again

Rakshith-R commented on internal/rbd/controllerserver_test.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2022 The Ceph-CSI Authors.
+Copyright 2025 The Ceph-CSI Authors.
```
```suggestion
Copyright 2024 The Ceph-CSI Authors.
```

Rakshith-R commented on internal/rbd/controllerserver_test.go:
```diff
@@ -16,71 +16,70 @@ limitations under the License.
+)
 
-func TestValidateStriping(t *testing.T) {
+func TestToCSISnapshot(t *testing.T) {
```
I think this should go into snapshot_test.go
and functions defined there should be in controllersrever_test.go ?

nixpanic replied:
A, yes, somehow the files were swapped. I wonder how I managed to do that :shrug:

Rakshith-R reviewed (dismissed):
Thanks !

nixpanic commented:
@mergifyio queue

nixpanic commented:
well, at least the change does what it needs to do, unfortunately it consistently hits this:
```
  I0218 13:41:57.114042       1 utils.go:266] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 GRPC call: /csi.v1.Controller/CreateSnapshot
  I0218 13:41:57.114081       1 utils.go:267] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 GRPC request: {"name":"snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7","parameters":{"clusterID":"730b65ad-7781-4098-b479-759fd21e9f17","csi.storage.k8s.io/volumesnapshot/name":"rbd-pvc-snapshot","csi.storage.k8s.io/volumesnapshot/namespace":"rbd-2355","csi.storage.k8s.io/volumesnapshotcontent/name":"snapcontent-e627b801-8de6-4f8f-b583-a4f9215cc0d7"},"secrets":"***stripped***","source_volume_id":"0001-0024-730b65ad-7781-4098-b479-759fd21e9f17-0000000000000004-ca406e9d-e195-4a08-8998-2a5c413b5449"}
  I0218 13:41:57.115160       1 omap.go:89] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 got omap values: (pool="replicapool", namespace="", name="csi.volume.ca406e9d-e195-4a08-8998-2a5c413b5449"): map[csi.imageid:151cd333273f csi.imagename:csi-vol-ca406e9d-e195-4a08-8998-2a5c413b5449 csi.volname:pvc-f0302f5b-ea76-4187-bde7-ffec8d519ced csi.volume.owner:rbd-2355]
  E0218 13:41:57.130825       1 omap.go:80] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 omap not found (pool="replicapool", namespace="", name="csi.snaps.default"): rados: ret=-2, No such file or directory
  I0218 13:41:57.174913       1 omap.go:159] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 set omap keys (pool="replicapool", namespace="", name="csi.snaps.default"): map[csi.snap.snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7:a5386ad7-2cf6-46ba-a19a-12718d8b6959])
  I0218 13:41:57.178656       1 omap.go:159] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 set omap keys (pool="replicapool", namespace="", name="csi.snap.a5386ad7-2cf6-46ba-a19a-12718d8b6959"): map[csi.imagename:csi-snap-a5386ad7-2cf6-46ba-a19a-12718d8b6959 csi.snapname:snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 csi.source:csi-vol-ca406e9d-e195-4a08-8998-2a5c413b5449 csi.volume.owner:rbd-2355])
  I0218 13:41:57.178681       1 rbd_journal.go:445] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 generated Volume ID (0001-0024-730b65ad-7781-4098-b479-759fd21e9f17-0000000000000004-a5386ad7-2cf6-46ba-a19a-12718d8b6959) and image name (csi-snap-a5386ad7-2cf6-46ba-a19a-12718d8b6959) for request name (snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7)
  I0218 13:41:57.178736       1 rbd_util.go:1455] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 rbd: snap create replicapool/csi-vol-ca406e9d-e195-4a08-8998-2a5c413b5449@csi-snap-a5386ad7-2cf6-46ba-a19a-12718d8b6959 using mon rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789
  I0218 13:41:58.092796       1 rbd_util.go:1493] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 rbd: clone replicapool/csi-vol-ca406e9d-e195-4a08-8998-2a5c413b5449@csi-snap-a5386ad7-2cf6-46ba-a19a-12718d8b6959 replicapool/csi-snap-a5386ad7-2cf6-46ba-a19a-12718d8b6959 (features: [layering deep-flatten]) using mon rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789
  I0218 13:41:58.092828       1 rbd_util.go:1603] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 setting image options on replicapool/csi-snap-a5386ad7-2cf6-46ba-a19a-12718d8b6959
  I0218 13:41:58.176001       1 rbd_util.go:1468] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 rbd: snap rm replicapool/csi-vol-ca406e9d-e195-4a08-8998-2a5c413b5449@csi-snap-a5386ad7-2cf6-46ba-a19a-12718d8b6959 using mon rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789
  I0218 13:41:58.206045       1 rbd_util.go:1455] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 rbd: snap create replicapool/csi-snap-a5386ad7-2cf6-46ba-a19a-12718d8b6959@csi-snap-a5386ad7-2cf6-46ba-a19a-12718d8b6959 using mon rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789
  I0218 13:41:59.145495       1 omap.go:159] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 set omap keys (pool="replicapool", namespace="", name="csi.snap.a5386ad7-2cf6-46ba-a19a-12718d8b6959"): map[csi.imageid:151ca2450b5d])
  I0218 13:41:59.409768       1 controllerserver.go:1242] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 Removing clone image "replicapool/csi-snap-a5386ad7-2cf6-46ba-a19a-12718d8b6959"
  I0218 13:41:59.410010       1 rbd_util.go:659] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 rbd: delete csi-snap-a5386ad7-2cf6-46ba-a19a-12718d8b6959 using mon rook-ceph-mon-a.rook-ceph.svc.cluster.local:6789, pool replicapool
  I0218 13:41:59.441775       1 rbd_util.go:705] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 rbd: adding task to remove image "replicapool/csi-snap-a5386ad7-2cf6-46ba-a19a-12718d8b6959" with id "151cd333273f" from trash
  E0218 13:41:59.445054       1 rbd_util.go:716] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 failed to add task to delete rbd image: replicapool/csi-snap-a5386ad7-2cf6-46ba-a19a-12718d8b6959, rados: ret=-2, No such file or directory: "[errno 2] RBD image not found (error retrieving image from trash)"
  E0218 13:41:59.445076       1 controllerserver.go:1245] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 failed to delete clone image "replicapool/csi-snap-a5386ad7-2cf6-46ba-a19a-12718d8b6959": rados: ret=-2, No such file or directory: "[errno 2] RBD image not found (error retrieving image from trash)"
  I0218 13:41:59.451414       1 omap.go:126] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 removed omap keys (pool="replicapool", namespace="", name="csi.snaps.default"): [csi.snap.snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7]
  E0218 13:41:59.451485       1 utils.go:271] ID: 35 Req-ID: snapshot-e627b801-8de6-4f8f-b583-a4f9215cc0d7 GRPC error: rpc error: code = Internal desc = "replicapool/csi-snap-a5386ad7-2cf6-46ba-a19a-12718d8b6959@" does not have a source-volume-id set
```
[logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.30/detail/mini-e2e-helm_k8s-1.30/503/pipeline)

A snapshot should have the `SourceVolumeId` pointing to the volume-id of the original volume?

@Rakshith-R or @Madhu-1 , what do you think?

Madhu-1 commented:
>A snapshot should have the SourceVolumeId pointing to the volume-id of the original volume?

yes that correct it need to be set as its a required field in the response.

nixpanic committed:
rbd: always include the `SourceVolumeID` when returning a Snapshot

`doSnapshotClone()` returns a new `rbdVolume` object from a temporary
snapshot. This conversion drops the `SourceVolumeID` attribute, as a
`rbdVolume` does not have that.

After converting the `rbdVolume` back to a `rbdSnapshot`, the
`SourceVolumeID` attribute can be set again, and the `ToCSI()` function
can create an appropriate CSI Snapshot struct.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.32

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1248,7 +1259,11 @@ func (cs *ControllerServer) CreateSnapshot(
 
-	csiSnap, err := vol.toSnapshot().ToCSI(ctx)
+	// FIXME: doSnapshotClone() returns a rbdVolume, some attributes may be missing?
+	snap := vol.toSnapshot()
```
@nixpanic i think it make sense to set this inside `toSnapshot` function

nixpanic replied:
@Madhu-1, that can not be done, as an `rbdVolume` does not have a `SourceVolumeID`, it is only part of `rbdSnapshot`. Maybe `doSnapshotClone()` should return an `rbdSnapshot`, or something else in the flow needs to be adjusted. I know that this isn't very clean, hence the **FIXME** comment.

Madhu-1 replied:
ha, i missed that one. Thanks

Rakshith-R replied:
> @Madhu-1, that can not be done, as an `rbdVolume` does not have a `SourceVolumeID`, it is only part of `rbdSnapshot`. Maybe `doSnapshotClone()` should return an `rbdSnapshot`, or something else in the flow needs to be adjusted. I know that this isn't very clean, hence the **FIXME** comment.

Can we please have an issue open for this too ?

nixpanic replied:
Done, see #5164

Rakshith-R replied:
Thanks

Rakshith-R reviewed (approved):
Thanks !

iPraveenParihar commented:
Thanks! @nixpanic

iPraveenParihar commented:
@Mergifyio queue

nixpanic commented:
/retest ci/centos/mini-e2e/k8s-1.31

nixpanic commented:
> /retest ci/centos/mini-e2e/k8s-1.31

@mergifyio requeue

Failed with:
```
[FAILED] backend images not matching kubernetes resource count,image count 1 kubernetes resource count 0
  backend image Info:
   [csi-vol-47dd54b3-94f8-4647-9e2a-196304d5302b]
   images information and status Pool: replicapool, Image: csi-vol-47dd54b3-94f8-4647-9e2a-196304d5302b, Info: {"name":"csi-vol-47dd54b3-94f8-4647-9e2a-196304d5302b","id":"16eb56e3021b","size":1073741824,"objects":256,"order":22,"object_size":4194304,"snapshot_count":0,"block_name_prefix":"rbd_data.16eb56e3021b","format":2,"features":["layering"],"op_features":[],"flags":[],"create_timestamp":"Thu Feb 20 07:12:17 2025","access_timestamp":"Thu Feb 20 07:12:17 2025","modify_timestamp":"Thu Feb 20 07:12:17 2025"}
  , Status: {"watchers":[]}
```
[logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.31/detail/mini-e2e_k8s-1.31/389/pipeline)

nixpanic commented:
@mergifyio requeue
